docs/architecture/adr-102-rpc-companion.md

```diff
@@ -0,0 +1,344 @@
+assume there's a `Block` structure, let's call it `v1`. If in the future there's a need to modify this structure that is not compatible with the
+previous data structure, then the database would support a `v2` schema for `ResultBlock` and an `index` table could determine the criteria on which
+data structure should be used for inserting or querying data.
+
+There's also a possibility that the data structure could be stored in parallel with the normalized data but in one blob, e.g. in a `jsonb` field
+that contains all the information needed to return the client, this could be an optimized way to serve the data but could add to the storage
+requirements. This could improve the performance for reading data from the database. An alternative approach to this problem that can offer
+similar results would be to use a "caching" layer.
+
+
+### RPC server instance
+
+The **RPC server instance** is a node that runs the RPC API process for the data companion. This server instance provide an RPC API (v1) with
+the same JSONRPC methods in the full node endpoint. The RPC Companion service will expose the same JSONRPC methods and will accept the same request types and
+return wire compatible responses (should match the same response as the equivalent full node RPC endpoint).
+
+The **RPC server instance**, when serving a particular request, retrieves the required data from the database in order to
+fulfill the request. The data should be serialized in a way that makes it wire compatible with the CometBFT RPC endpoint.
+
+Identical requests should return idempotent responses, no side effects should cause the RPC service to return different responses.
+
+The RPC service endpoints should be exposes through an external load-balancer service such as Cloudflare or AWS ELB, or
+a server running its own load balancer mechanism (e.g. nginx).
+
+The RPC clients should make requests to the **RPC Companion** server instances through this load balancer.
+
+The **RPC server instance** will also implement logic for the proxied requests to the full node. It should properly handle the proxy requests and responses from the full node.
+
+The **RPC Companion** should support the `https` protocol in order to support secure endpoints access. It's recommended that
+the `https` support is provided by the load balancer but in case there's a single **RPC Companion** server instance, having an
+option to use `https` is important and it's backwards compatible with the existing CometBFT RPC that supports that.
+
+
+## Consequences
+
+### Positive
+
+- Alternative and optional **RPC Companion** that is more scalable and reliable with a higher query throughput.
+- Less backpressure on the full node that is running consensus.
+- Possibility for future additional endpoints (v2).
+- Allow users to create better and faster indexers and analytics solutions.
+
+### Negative
+
+- Additional infrastructure complexity to set up and maintain.
+- Additional infrastructure cost.
```

Comment:
This might actually not be only a cost because it should save on some storage (especially if we port the indexes), and CPU cost when querying the full nodes whose resources we were taking up until now. Of course it will cost more but maybe not *that* much more.

Reply:
Maybe I should be more specific on infrastructure cost. I was thinking mostly the whole setup if using multiple servers for the RPC Companion service, also the database depending on the setup (master/replica), load balancer, etc. But agree, some of the cost might be offset if the data is properly stored and normalized.